eos-paygd/main.c

```diff
@@ -194,24 +240,52 @@ main (int   argc,
       watchdog_id = g_timeout_add_seconds_full (G_PRIORITY_HIGH, 10, ping_watchdog, NULL, NULL);
       g_assert (watchdog_id > 0);
 
-      /* Let systemd know it's okay to proceed to pivot to the final root */
-      sd_notify_ret = sd_notify (0, "READY=1");
+      /* Here be dragons:
+       * We're currently in the initramfs root directory, and systemd is putting
+       * all the useful bits of the system into /sysroot in preparation for the
+       * root pivot. After the root pivot, it will delete everything in the
+       * initramfs root, which will leave us in limbo forever.  We need to
+       * chroot() out of this into sysroot.
+       *
+       * However, once we do that, we no longer have access to the socket we
+       * need to write to to inform systemd we're ready for it to do the root
+       * pivot.
+       *
+       * What we do is chdir() into the directory $NOTIFY_SOCKET resides in,
+       * chroot() into /sysroot, then use our own variant of sd_notify() that
+       * doesn't refuse to use relative paths to notify systemd we're ready.
+       *
+       * After that, we can chdir() into our new / and begin trying to connect
+       * to the dbus socket. */
+
+      sd_socket_env = g_getenv("NOTIFY_SOCKET");
+      if (!sd_socket_env)
+        {
+          /* If we can't notify systemd that we're ready to move on, it will
+           * timeout in 90 seconds and shutdown, might as well just exit. */
+          g_printerr ("%s: NOTIFY_SOCKET not set.\n", argv[0]);
+          return SD_NOTIFY_FAILURE_EXIT_CODE;
```

Comment:
By returning early we're skipping the `allow_writing_to_boot_partition (FALSE);` below. I guess that's not an issue; the boot partition is writable while untrusted userspace code is running anyway. Just wanted to make sure we're aware.

Reply:
Same situation exists for the watchdog timer, but I think at this point an exit with failure code is going to result in an immediate shutdown while we're still in early initrd?

Reply:
You're right we do have `FailureAction=poweroff`